vlang: 2022.20 -> 2023.01

COLAMAroro opened the pull request:
This is my first package on nixpkgs, fell free to raise any issues, especially concerning the form :smiley: 

###### Description of changes

- Bump from 2022.20 to 2023.01
- Make optionals dependencies optional
- Bypass autodetection on Makefile (which would build for wrong system)
- Add boehmgc as a static lib and patch it in sources
- Patches unknown shebangs

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

COLAMAroro committed:
vlang: 2022.20 -> 2023.01

Changes on this version bump:
- Bump from 2022.20 to 2023.01
- Remove optional dependencies (like freetype or openssl)
- Bypass autodetection on Makefile (which would build for wrong system)
- Add boehmgc as a static lib and patch it in sources
- Patches unknown shebangs
- Add a major comment on those removed optional deps

COLAMAroro commented:
I've checked if this would produce ecosystems through `grep -irn "vlang" --exclude-dir=.git`. There is only the `vlang` package itself, and the vim language server (which references their own github repo, not the `vlang` package), so no risk of breaking anything else in the nixpkgs

https://github.com/NixOS/nixpkgs/blob/05c56f84c67e9df67046a70898efd3675e6f9ce0/pkgs/applications/editors/vim/plugins/nvim-treesitter/generated.nix#L1497

```grep
pkgs/top-level/all-packages.nix:15710:  vlang = callPackage ../development/compilers/vlang { };
pkgs/applications/editors/vim/plugins/nvim-treesitter/generated.nix:1501:      owner = "vlang";
pkgs/applications/editors/vim/plugins/nvim-treesitter/generated.nix:1507:    meta.homepage = "https://github.com/vlang/vls";
pkgs/development/compilers/vlang/default.nix:15:  pname = "vlang";
pkgs/development/compilers/vlang/default.nix:19:    owner = "vlang";
pkgs/development/compilers/vlang/default.nix:27:    owner = "vlang";
pkgs/development/compilers/vlang/default.nix:35:    owner = "vlang";
pkgs/development/compilers/vlang/default.nix:120:      homepage = "https://vlang.io/";
```

COLAMAroro commented:
`nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"` and `nixpkgs-review pr 208787` built without issue

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1670

SuperSandro2000 commented on pkgs/development/compilers/vlang/default.nix:
```diff
@@ -39,6 +47,18 @@ stdenv.mkDerivation rec {
     export HOME=$(mktemp -d)
+    # We need to set the target system, because the auto-detection doesn't work that great.
+    export VFLAGS="-os ${targetSystem} -showcc -skip-unused"
+    # We also need gc.h in the include path.
+    export CFLAGS="-I${staticBoehmgc.dev}/include"
+    export LDFLAGS="-L${staticBoehmgc}/lib"
```
Does adding the package to buildInputs work?

COLAMAroro replied:
Fixed in https://github.com/COLAMAroro/nixpkgs/blob/dcaed354bd52104677eb30460088e627ce74e465/pkgs/development/compilers/vlang/default.nix#L52

SuperSandro2000 commented on pkgs/development/compilers/vlang/default.nix:
```diff
@@ -39,6 +47,18 @@ stdenv.mkDerivation rec {
+    export CFLAGS="-I${staticBoehmgc.dev}/include"
+    export LDFLAGS="-L${staticBoehmgc}/lib"
+
+    # Monkeypatch the source code like an animal to use BoehmGC static files,
+    # instead of the "thirdparty" one.
+    # We should replace "#flag @VEXEROOT/thirdparty/tcc/lib/libgc.a"
+    # to "#flag ${staticBoehmgc}/lib/libgc.a"
+    # in file vlib/builtin/builtin_d_gcboehm.c.v
```
```suggestion
    # patch the code to use our static compiled BoehmGC
```

SuperSandro2000 commented on pkgs/development/compilers/vlang/default.nix:
```diff
@@ -76,12 +96,32 @@ stdenv.mkDerivation rec {
-    platforms = platforms.all;
-  };
+  preFixup = ''
+    # We need to patch the shebang of file "lib/cmd/tools/git_pre_commit_hook.vsh"
+    # and file "lib/vlib/v/tests/script_with_no_extension"
+    # Going from "#!/usr/bin/env -S v -raw-vsh-tmp-prefix tmp"
+    # To "#!$out/v -raw-vsh-tmp-prefix tmp"
```
```suggestion
    # We need to patch the shebangs to point to our v
```

we don't need to repeat the literal code. Also can this be done via patchShebangs?

COLAMAroro replied:
The default shebang patcher didn't recognize those shebangs (probably due to the `#!/usr/bin/env -S`, so I went this route instead. While it would probably work if patched with `patchShebangs`, the real reason the V developers decided to use `#!/usr/bin/env` is that V could be installed from anywhere. It made more sense to me to just fix it to our actual V instead

COLAMAroro replied:
Change suggestion fixed in https://github.com/COLAMAroro/nixpkgs/blob/dcaed354bd52104677eb30460088e627ce74e465/pkgs/development/compilers/vlang/default.nix#L118

SuperSandro2000 commented on pkgs/development/compilers/vlang/default.nix:
```diff
@@ -76,12 +96,32 @@ stdenv.mkDerivation rec {
+  '';
+
+  /* Details concerning features:
+    If you want to use `-compress`, you need to add `upx` in your build inputs.
+    If you want to build x11 apps (like examples/2048.v) you will need:
+    - `xorg.libX11.dev`
+    - `xorg.libXi.dev`
+    - `xorg.libXcursor.dev`
+    - `libGL.dev`
+    - `freetype`
```
This should be toggable via a boolean instead.

COLAMAroro replied:
I should do so by adding a `buildWithExtras` as an input of the function, to make it available through `override` ?

COLAMAroro replied:
Fixed in https://github.com/COLAMAroro/nixpkgs/blob/dcaed354bd52104677eb30460088e627ce74e465/pkgs/development/compilers/vlang/default.nix#L54

SuperSandro2000 commented on pkgs/development/compilers/vlang/default.nix:
```diff
@@ -76,12 +96,32 @@ stdenv.mkDerivation rec {
+    - `xorg.libXcursor.dev`
+    - `libGL.dev`
+    - `freetype`
+    If you want to build web apps (using `net.http` or `net.websocket`) you will need `openssl`.
```
to link against or does v need to be build with that?

COLAMAroro replied:
It is not needed for V itself to build. However, because V basically transpiles to C, the actual C compiler needs those dependencies if you decide to use the feature.

SuperSandro2000 commented on pkgs/development/compilers/vlang/default.nix:
```diff
@@ -76,12 +96,32 @@ stdenv.mkDerivation rec {
+    - `freetype`
+    If you want to build web apps (using `net.http` or `net.websocket`) you will need `openssl`.
+  */
+  meta = with lib;
+    {
+      homepage = "https://vlang.io/";
+      description = "Simple, fast, safe, compiled language for developing maintainable software";
+      license = licenses.mit;
+      maintainers = with maintainers; [ Madouura ];
+      mainProgram = "v";
+      platforms = platforms.all;
+    };
```
```suggestion
  meta = with lib; {
    homepage = "https://vlang.io/";
    description = "Simple, fast, safe, compiled language for developing maintainable software";
    license = licenses.mit;
    maintainers = with maintainers; [ Madouura ];
    mainProgram = "v";
    platforms = platforms.unix;
  };
```

SuperSandro2000 commented on pkgs/development/compilers/vlang/default.nix:
```diff
@@ -1,35 +1,43 @@
     repo = "v";
     rev = version;
-    sha256 = "1isbyfs98bdbm2qjf7q4bqbpsmdiqlavn3gznwr12bkvhnsf4j3x";
+    sha256 = "G/8SF0g/YY0kmDdVSTzuI5ksrrwaFuhxZILIIGRdbyw=";
```
```suggestion
    sha256 = "sha256-baHN9m2vrjPH0cIf7WoHcH5JH1HUFNpabQklY0jjKgQ=";
```

COLAMAroro committed:
Apply low-hanging suggestions from code review

- Changes sha256 format
- Change BoehmGC comment to be less verbose
- Change brackets location for `meta`
- Specify in `meta` that the platforms are unix only

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

COLAMAroro committed:
Apply other suggestions from code review

- Enable extra options through attributs
  - Use `buildWithUpx` for the `-compress` flag
  - Use `buildWithX11` for the `X11` support
  - Use `buildWithWeb` for the web support
  - It doesn't influence the compiler itself, only programs built with V
- Use `buildInputs` for BoehmGC (instead of manually settings `CFLAGS`
- Less verbose comment for shebang patching

COLAMAroro commented:
Closing PR, will open another cleaner one when 2023-03 releases
